**Objective:** Develop a data pipeline to collect, analyze, and visualize food waste data from various sources. Implement predictive models to forecast future food waste trends, providing insights to stakeholders for informed decision-making.

1. Identify required features for you analysis and gather data

# Data Understanding

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("D:/Sanjana/Projects/resources/food_waste_FAOU.csv")

In [3]:
df.head()

m49_code  country region  cpc_code     commodity  year  loss_percentage  \
0       104  Myanmar    NaN  23161.02  Rice, milled  2015             1.78   
1       104  Myanmar    NaN  23161.02  Rice, milled  2015            11.77   
2       104  Myanmar    NaN  23161.02  Rice, milled  2015             5.88   
3       104  Myanmar    NaN  23161.02  Rice, milled  2015             3.57   
4       104  Myanmar    NaN  23161.02  Rice, milled  2015            17.65   

  loss_percentage_original loss_quantity activity food_supply_stage  \
0                    1.78%      26.12kgs  Storage           Storage   
1                   11.77%      88.18kgs  Storage           Storage   
2                    5.88%      44.09kgs  Storage           Storage   
3                    3.57%      52.24kgs  Storage           Storage   
4                   17.65%     132.27kgs  Storage           Storage   

                        treatment cause_of_loss sample_size  \
0  30 days storage, with trapping       Rodents         NaN   
1    60 days storage, no trapping       Rodents         NaN   
2    30 days storage, no trapping       Rodents         NaN   
3  60 days storage, with trapping       Rodents         NaN   
4    90 days storage, no trapping       Rodents         NaN   

  method_data_collection                                          reference  \
0  Controlled Experiment  Dr Steven Belmain (2015), context post-harvest...   
1  Controlled Experiment  Dr Steven Belmain (2015), context post-harvest...   
2  Controlled Experiment  Dr Steven Belmain (2015), context post-harvest...   
3  Controlled Experiment  Dr Steven Belmain (2015), context post-harvest...   
4  Controlled Experiment  Dr Steven Belmain (2015), context post-harvest...   

   url                                        notes  
0  NaN   Reference has been generated automatically  
1  NaN   Reference has been generated automatically  
2  NaN   Reference has been generated automatically  
3  NaN   Reference has been generated automatically  
4  NaN   Reference has been generated automatically

In [41]:
print("List of recorded activities leading to food wastage: ")
activities = df["activity"].dropna()

unique_activities = set()

for activity in activities:
    split_activities = activity.split(",")  
    cleaned_activities = [a.strip().lower() for a in split_activities]  
    unique_activities.update(cleaned_activities) 

unique_activities

List of recorded activities leading to food wastage: 


{'bagging',
 'blanching',
 'bulking',
 'bundling',
 'cleaning',
 'collection',
 'consumption',
 'distribution',
 'drying',
 'farm',
 'field',
 'freezing',
 'grading',
 'handling',
 'harvesting',
 'layering',
 'lifting',
 'manufacturing',
 'marketing',
 'milling',
 'packaging',
 'peeling',
 'piling',
 'processing',
 'retailing',
 'ripening',
 'shelling',
 'sorting',
 'stacking',
 'storage',
 'threshing',
 'trading',
 'transportation',
 'washing',
 'wholesale',
 'winnowing'}

In [8]:
print("Recorded countries for food wastage by Food and Agricultuaral Organization of the United Nations: ", len(df["country"].unique()))
df["country"].unique()

Recorded countries for food wastage by Food and Agricultuaral Organization of the United Nations:  105


array(['Myanmar', 'Burundi', 'Cambodia', 'Cameroon', 'Canada',
       'Sri Lanka', 'Chad', 'Europe', 'Chile', 'China', 'Colombia',
       'Democratic Republic of the Congo', 'Costa Rica', 'Benin',
       'Denmark', 'Northern America', 'Ecuador', 'El Salvador',
       'Ethiopia', 'Eritrea', 'Angola', 'Fiji', 'Finland', 'France',
       'Gabon', 'Gambia', 'Palestine', 'Ghana', 'Azerbaijan', 'Argentina',
       'Guatemala', 'Guinea', 'Guyana', 'Honduras', 'India', 'Australia',
       'Indonesia', 'Iran (Islamic Republic of)', 'Italy',
       "Côte d'Ivoire", 'Kazakhstan', 'Jordan', 'Kenya',
       "Democratic People's Republic of Korea", 'Republic of Korea',
       "Lao People's Democratic Republic", 'Lebanon', 'Lesotho',
       'Liberia', 'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mexico',
       'Bangladesh', 'Morocco', 'Mozambique', 'Armenia', 'Namibia',
       'Nepal', 'Australia and New Zealand', 'Nicaragua', 'Niger',
       'Nigeria', 'Norway', 'Pakistan', 'Panama', 'Paraguay', 

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11834 entries, 0 to 11833
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   m49_code                  11834 non-null  int64  
 1   country                   11834 non-null  object 
 2   region                    600 non-null    object 
 3   cpc_code                  11834 non-null  object 
 4   commodity                 11834 non-null  object 
 5   year                      11834 non-null  int64  
 6   loss_percentage           11834 non-null  float64
 7   loss_percentage_original  11834 non-null  object 
 8   loss_quantity             265 non-null    object 
 9   activity                  10589 non-null  object 
 10  food_supply_stage         10288 non-null  object 
 11  treatment                 750 non-null    object 
 12  cause_of_loss             567 non-null    object 
 13  sample_size               612 non-null    object 
 14  method

## Data Cleaning

### Null values

In [47]:
# Percentage of null values
df["notes"].isnull().sum()/df["notes"].size
# We can drop this column, high percentage of null values

0.8484874091600473

In [54]:
usable_df = df.drop(columns = ["notes","url","reference","method_data_collection", "sample_size"])

In [55]:
usable_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11834 entries, 0 to 11833
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   m49_code                  11834 non-null  int64  
 1   country                   11834 non-null  object 
 2   region                    600 non-null    object 
 3   cpc_code                  11834 non-null  object 
 4   commodity                 11834 non-null  object 
 5   year                      11834 non-null  int64  
 6   loss_percentage           11834 non-null  float64
 7   loss_percentage_original  11834 non-null  object 
 8   loss_quantity             265 non-null    object 
 9   activity                  10589 non-null  object 
 10  food_supply_stage         10288 non-null  object 
 11  treatment                 750 non-null    object 
 12  cause_of_loss             567 non-null    object 
dtypes: float64(1), int64(2), object(10)
memory usage: 1.2+ MB


In [58]:
print((usable_df["region"].isnull().sum())/usable_df.size)
print((usable_df["treatment"].isnull().sum())/usable_df.size)
# we can include these columns

0.07302297162023375
0.07204794529452295


### Standardize

In [75]:
usable_df["loss_quantity"].unique()

array(['26.12kgs', '88.18kgs', '44.09kgs', '52.24kgs', '132.27kgs',
       '78.36kgs', nan, '348 tonnes', '16.90kgs', '77.74kgs', '27.99kgs',
       '31.77kgs', '0 tons', '0.83 KG', '1498-5926 KG', '541-2893 KG',
       '4.07 quintal', '2.5 quintal', '3.34 quintal', '9.31 quintal',
       '11 - 14 M kg/year', '33 - 43 M kg/year', '67306', '1332657',
       '121151', '16911', '12176', '168265', '184230', '608', '1018',
       '50526', '28842', '4511', '1021', '70942', '874', '9668', '262582',
       '609', '35020', '32496', '1117', '45414', '26813', '4587', '1152',
       '67714', '854', '9272', '232567', '546', '29954', '893', '40437',
       '13294', '3880', '1030', '61911', '806', '8601', '242532', '592',
       '30203', '774', '40577', '12676', '3495', '877', '60946', '402',
       '8575', '260753', '590', '33363', '628', '34391', '11319', '3310',
       '704', '58502', '355', '6646', '189425', '683', '35022', '405',
       '33851', '10038', '3569', '49101', '687', '55938', '347', '

In [89]:
usable_df[usable_df["loss_quantity"].notnull()][["loss_quantity","loss_percentage"]].head(30)

loss_quantity  loss_percentage
0              26.12kgs           1.7800
1              88.18kgs          11.7700
2              44.09kgs           5.8800
3              52.24kgs           3.5700
4             132.27kgs          17.6500
5              78.36kgs           5.3500
273          348 tonnes          38.0000
692            16.90kgs           1.8600
721            77.74kgs           2.4100
747            27.99kgs           2.0000
765            31.77kgs           1.7500
786              0 tons           4.7000
787              0 tons          11.0000
788              0 tons           6.5000
789              0 tons           1.1000
790              0 tons           3.1000
791              0 tons           1.7500
792              0 tons           3.0000
1317            0.83 KG           2.4500
1318       1498-5926 KG          10.4650
1319        541-2893 KG           0.9305
1623       4.07 quintal           1.9200
1624        2.5 quintal           0.6500
1625       3.34 quintal           0.5900
1626       9.31 quintal          21.7200
2411  11 - 14 M kg/year           2.2500
2412  33 - 43 M kg/year           3.0000
2928              67306           1.0000
2929            1332657          20.0000
2930             121151           5.0000

In [126]:
# checking certain loss quantity values that are too high
usable_df[usable_df["loss_quantity"].str.contains("mt|M tonnes|M|m", na=False)]

m49_code      country region  cpc_code  \
2411       246      Finland    NaN  21111.01   
2412       246      Finland    NaN     22290   
7105       566      Nigeria    NaN  01520.01   
7248       608  Philippines    NaN      0112   
7249       608  Philippines    NaN      0113   
7250       608  Philippines    NaN     01312   

                                           commodity  year  loss_percentage  \
2411  Meat of cattle with the bone, fresh or chilled  2014             2.25   
2412                           Dairy products n.e.c.  2014             3.00   
7105                                  Cassava, fresh  2013             8.51   
7248                                    Maize (corn)  2022            14.69   
7249                                            Rice  2022            18.10   
7250                                         Bananas  2022            20.05   

     loss_percentage_original      loss_quantity       activity  \
2411                 2 - 2.5%  11 - 14 M kg/year  Manufacturing   
2412                       3%  33 - 43 M kg/year  Manufacturing   
7105                     8.51       3,191,599 mt            NaN   
7248                   14.69%     0.246 M tonnes            NaN   
7249                   18.10%       2.3 M tonnes            NaN   
7250                   20.05%     0.854 M tonnes            NaN   

       food_supply_stage treatment    cause_of_loss  
2411          Processing       NaN              NaN  
2412          Processing       NaN              NaN  
7105                Farm       NaN  Physical Losses  
7248  Whole supply chain       NaN              NaN  
7249  Whole supply chain       NaN              NaN  
7250  Whole supply chain       NaN              NaN

Nigeria,01520.01, Cassava, fresh,2013, loss_quantity = 3,191,599 mt	
seemed like a strangely huge loss, but according to available records, Nigeria's cassava output was 35 million tonnes in 2012 and was expected to rise to 95 million tonnes in 2013.

A loss of approximately 3.19 million mt represents about 3.36% of the 95 million mt total production in 2013, data shows there being a loos_percentage of 8.51%.

Physical Losses refers to the actual quantity of produce lost due to factors like spoilage, pests, diseases, or post-harvest handling issues.

And given that data was collected from the UN, the record has high plausibility therefore we won't discard it.

source: https://www.freshplaza.com/north-america/article/2116208/nigeria-cassava-production-expected-to-have-doubled-in-2013/

In [109]:
# standarize values

# 1 tonne = 1000 kg
# 1 quintal = 100 kg
# 1 mt = 1,000,000,000 kg

import re
for val in usable_df["loss_quantity"]:
    if re.search(r"kg ha-1", val):
        val = val.replace(val,np.nan)
    elif re.search(r"[ton|tons|tonnes|tonne]", val):
        
        print(val)
    elif re.search(r"[Kk][Gg][Ss]*", val):
        print(val)
    elif re.search(r"quintal", val):
        print(val)
    elif re.search(r"M tonnes|mt",val):
        print(val)
        

26.12kgs


In [117]:
np.nan

nan